zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/versioning/VersionsServiceImpl.java

```diff
@@ -0,0 +1,177 @@
+     * Otherwise return an empyt optional.
+     */
+    @Override
+    public Optional<String> getVersionNameFromURI(String uri) {
+        Optional<String> result = Optional.empty();
+
+        if (isVersionedURI(uri)) {
+            Matcher matcher = VERSION_DIR_PATTERN.matcher(uri);
+
+            if (matcher.find()) {
+                result = Optional.of(matcher.group(0).replace(VERSION_URI, ""));
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Return the version number of a uri if the input value is a valid version uri. A valid version uri is a non
+     * null, non empty string containing the substring matching the regex "/previous/v\\d+".
+     * Example: "/a/b/c/previous/v1/data.json" will return 1
+     *
+     * @param uri the input proccess.
+     * @return the version number an int if the input is valid. Returns the a default of -1 for all invalid inputs.
+     */
+    @Override
+    public int getVersionNumberFromURI(String uri) {
+        Integer version = -1;
+        Optional<String> vOpt = getVersionNameFromURI(uri);
+
+        if (vOpt.isPresent()) {
+            version = Integer.valueOf(vOpt.get().replaceFirst("v", ""));
+        }
+
+        return version;
+    }
+
+    @Override
+    public Comparator<Version> versionComparator() {
+        return (Version v1, Version v2) -> {
+            Integer foo = getVersionNumberFromURI(v1.getUri().toString());
+            Integer bar = getVersionNumberFromURI(v2.getUri().toString());
+            return foo.compareTo(bar);
+        };
+    }
+
+    @Override
+    public boolean isVersionDir(File file) {
+        return file != null && file.isDirectory() && VALID_VERSION_DIR_PATTERN.matcher(file.getName()).matches();
+    }
+
+
+    @Override
+    public void verifyCollectionDatasets(ZebedeeReader cmsReader, Collection collection,
+                                         CollectionReader reader, Session session)
+            throws ZebedeeException, IOException, VersionNotFoundException {
+
+        List<Dataset> datasets = getDatasetsInCollection(reader);
+
+        if (datasets != null && !datasets.isEmpty()) {
+            verifyDatasetVersions(cmsReader, collection, session, datasets);
+
+            info().collectionID(collection)
+                    .user(session)
+                    .data("datasets",
+                            datasets.stream()
+                                    .map(ds -> ds.getDescription().getTitle())
+                                    .collect(Collectors.toList()))
+                    .log("collection legacy datasets verified successfully");
+        }
+    }
+
+    List<Dataset> getDatasetsInCollection(CollectionReader reader) throws ZebedeeException, IOException {
+        List<Dataset> datasets = new ArrayList<>();
+
+        for (String uri : reader.getReviewed().listUris()) {
+
+            if (isDataJson(uri) && !isVersionedURI(uri)) {
```

Comment:
consider filtering by pages with `/datasets/` in the URI. That way it will not load the content for all data.json files in the collection. This could be quite expensive for large collections